Fix non-determinism in debuginfo

ilovepi opened the pull request:
Assignment tracking iterates over a SmallSet when adding metadata, which eventually results in debug metadata being added to the module in non-deterministic order.

As reported in #63921, we saw some cases where DWARF DebugLoc entries could have their order reversed, even though there was no functional difference.

This patch replaces the `SmallSet` with a `SmallVector`, and adds the required `DenseMapInfo` specialization to make the ordering deterministic.

Fixes #63921

ilovepi committed:
Fix non-determinism in debuginfo

Assignment tracking iterates over a SmallSet when adding metadata,
which eventually results in debug metadata being added to the module in
non-deterministic order.

As reported in #63921, we saw some cases where DWARF DebugLoc entries
could have their order reversed, even though there was no functional
difference.

This patch replaces the SmallSet with a SmallVector, and adds the
required DenseMapInfo specialization.

Fixes: #63921

ilovepi commented:
I tried to think of a way to test this, but since its a non-determinism issue, the only way I was reliably able to trigger this was to LD_PRELOAD a different default allocator, which isn't something I think we can really do. If there is a good way to add a test for this, I'm happy to add one.

fangism commented on llvm/include/llvm/IR/DebugInfo.h:
```diff
@@ -263,7 +265,8 @@ struct VarRecord {
 /// passed in with sret & byval parameters).
-using StorageToVarsMap = DenseMap<const AllocaInst *, SmallSet<VarRecord, 2>>;
+using StorageToVarsMap =
+    DenseMap<const AllocaInst *, SmallSetVector<VarRecord, 2>>;
```
Can you add a comment explaining the choice of `SmallSetVector` and nondeterminism?  (Just so no one feels tempted to change it back.)

ilovepi replied:
I'm not sure that is necessary. SmallSetVector is a type whose main use is to prevent non-deterministic output.

zero9178 commented on llvm/include/llvm/IR/DebugInfo.h:
```diff
@@ -314,6 +317,25 @@ class AssignmentTrackingPass : public PassInfoMixin<AssignmentTrackingPass> {
 bool isAssignmentTrackingEnabled(const Module &M);
+
+template <> struct DenseMapInfo<at::VarRecord> {
+  static inline at::VarRecord getEmptyKey() {
+    return at::VarRecord{nullptr, nullptr};
+  }
+
+  static inline at::VarRecord getTombstoneKey() {
+    return at::VarRecord{nullptr, nullptr};
+  }
```
The tombstone key and empty key mustn't be equal as this breaks the implementation of `DenseMap`. It can lead to accidently finding a member on lookup when it shouldn't or not finding a member when it should.

What you can do is use e.g. `DenseMapInfo<DILocalVariable*>::getTombstoneKey()` as a value here and the corresponding empty key in the other.

ilovepi replied:
Thank you. I was 90% sure I was doing this wrong.

ilovepi committed:
[llvm][debuginfo] Use DenseMapInfo keys from component types

The naive implementation would break DenseMap, and yield wrong results.
Instead use the Tobstone and Empty Keys from the field types of
VarRecord.

OCHyams commented:
Thanks for hunting this down. LGTM with an inline comment. I'll let @fangism accept when they're happy.

OCHyams reviewed (commented):
(apparently I didn't submit my comments)

ilovepi committed:
Replace VarRecord implementation with std::pair as suggested

fangism reviewed (approved):
I'm not familiar enough with the code base for my approval to be worth anything, but nevertheless I am thrilled this is getting fixed!

I only wish there was a better way to test for determinism in isolation.

ilovepi commented:
huh, not sure why `check-llvm` was passing for me locally, since it looks like the change w/ `std::pair` is just broken. 

I'm probably going to go back to the DenseMapInfo Specialization, unless this ends up being simple to fix.

ilovepi committed:
Restore VarRecord, to minimize changes across the codebase.

ilovepi commented on llvm/include/llvm/IR/DebugInfo.h:
```diff
@@ -314,6 +317,27 @@ class AssignmentTrackingPass : public PassInfoMixin<AssignmentTrackingPass> {
+    return A == B;
+  }
+};
+
```
So I ended up going back to the `Varrecord` w/ the `DenseMapInfo` specialization, and placed it as close as possible to the class def. On the downside, it splits the `at` namespace.
